Refactor ProjectReferences in AppHost

eerhardt opened the pull request:
At a high-level there are 2 problems with how we are using ProjectReference to reference the individual "service application" projects in the AppHost.

1. The AppHost may not have a compatible TFM with the referenced application project. For example if the referenced application project was targeting `net8-windows`.
2. When we are generating IServiceMetadata code, we are using ReferencePathWithRefAssemblies items, which means if the ProjectReference has ReferenceOutputAssembly="false" on it, no IServiceMetadata code will be generated.

To resolve these issues, we are taking the following approach:

- By default ProjectReferences in the AppHost will mean a reference to a "service application", and not a regular project reference. If an AppHost project truly wants a ProjectReference to another project, it can be marked as `<ProjectReference Include="..." AppHostLibrary="true" />`.
- ProjectReferences to service applications will be marked `ReferenceOutputAssembly=false`, `SkipGetTargetFrameworkProperties=true`, `ExcludeAssets=all`. This will cause the AppHost to not reference the service app projects.
- Refactor the IServiceMetadata code gen to act directly on ProjectReference items.

Fix https://github.com/dotnet/aspire/issues/1074
Fix https://github.com/dotnet/aspire/issues/1563
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/1726)

eerhardt committed:
Refactor ProjectReferences in AppHost

At a high-level there are 2 problems with how we are using ProjectReference to reference the individual "service application" projects in the AppHost.

1. The AppHost may not have a compatible TFM with the referenced application project. For example if the referenced application project was targeting `net8-windows`.
2. When we are generating IServiceMetadata code, we are using ReferencePathWithRefAssemblies items, which means if the ProjectReference has ReferenceOutputAssembly="false" on it, no IServiceMetadata code will be generated.

To resolve these issues, we are taking the following approach:

- By default ProjectReferences in the AppHost will mean a reference to a "service application", and not a regular project reference. If an AppHost project truly wants a ProjectReference to another project, it can be marked as `<ProjectReference Include="..." AppHostLibrary="true" />`.
- ProjectReferences to service applications will be marked `ReferenceOutputAssembly=false`, `SkipGetTargetFrameworkProperties=true`, `ExcludeAssets=all`. This will cause the AppHost to not reference the service app projects.
- Refactor the IServiceMetadata code gen to act directly on ProjectReference items.

Fix #1074
Fix #1563

eerhardt committed:
Move ProjectReference defaulting logic into the SDK targets so it is respected by NuGet Restore

baronfel commented:
Looking forward to checking this out! One scenario I wanted to make sure we could handle is an Aspire AppHost referencing a self contained app - that currently breaks today.

eerhardt committed:
Fix test AppHost projects to also import the SDK.targets

eerhardt commented:
> One scenario I wanted to make sure we could handle is an Aspire AppHost referencing a self contained app - that currently breaks today.

Do you have repro steps I can try?

I took [eshop](https://github.com/dotnet/eShop), made one of the projects `<SelfContained>true</SelfContained>` and it still works.

eerhardt committed:
Suppress importing the orchestration SDK targets in test AppHosts, so the projects build even if the workload is not installed.

DamianEdwards commented on src/Aspire.Hosting.Sdk/SDK/Sdk.targets:
```diff
@@ -20,7 +18,21 @@
+  -->
+  <ItemGroup Condition="'$(IsAspireHost)' == 'true'">
+    
+    <ProjectReference Update="@(ProjectReference)">
```
Should we add a global opt-out property for this behavior? e.g. `EnableAspireProjectReferenceProcessing`

davidfowl replied:
Can you think of a good reason why?

DamianEdwards replied:
Still slightly uneasy about taking over ProjectReferences and it's common for MSBuild defaults to have an opt-out. But we can wait for feedback.

DamianEdwards commented on src/Aspire.Hosting/build/Aspire.Hosting.targets:
```diff
@@ -13,17 +13,14 @@
+      <_ServiceAppProject Include="@(ProjectReference->WithMetadataValue('_IsServiceApp', 'true'))" />
+
+      <ServiceMetadataSource Include="@(_ServiceAppProject)" Condition="'@(_ServiceAppProject)' != ''">
+        <ClassName Condition="%(_ServiceAppProject.ServiceNameOverride) == ''">$([System.IO.Path]::GetFileNameWithoutExtension(%(_ServiceAppProject.Identity)).Replace(".", "_").Replace("-","_").Replace(" ","_"))</ClassName>
```
Should we update this from `ServiceName` to `ResourceName` now?

DamianEdwards replied:
Or maybe just `AspireResourceName` and that can be used as the opt-out too by setting it to empty?

eerhardt replied:
What about `ProjectsClassName`? Or even `ProjectsName`?

I don't like calling it "ResourceName" because that is what we call the string parameter that pass in to `AddProject`, which could be a different name than this name.

> and that can be used as the opt-out too by setting it to empty?

In MSBuild, I haven't seen setting something to empty and not setting it being different behavior. Usually checking for `== ''` means "it isn't set".

baronfel replied:
For Items we do have the ['HasMetadata' item function](https://learn.microsoft.com/en-us/visualstudio/msbuild/item-functions?view=vs-2022), but aside from this it can be hard to distinguish between missing and empty.

eerhardt replied:
I don't like setting this metadata to empty to mean something completely different than what the metadata is intended for.

* `<ProjectReference Include=.. />` - means this is a service app and don't reference it, use the default name for the Projects class name
* `<ProjectReference Include=.. AspireResourceName="foo" />` - means this is a service app and don't reference it, use "foo" for the Projects class name
* `<ProjectReference Include=.. AspireResourceName="" />` - means this isn't a service app and should be referenced

I do think we should rename `ServiceNameOverride`, but I don't know of a good name. If people think `AspireResourceName` is OK, I can rename it. I just have the reservations I listed above.

DamianEdwards commented on tests/Aspire.Hosting.Tests/Aspire.Hosting.Tests.csproj:
```diff
@@ -8,11 +8,18 @@
-    <ProjectReference Include="..\..\src\Aspire.Hosting.Azure\Aspire.Hosting.Azure.csproj" />
-    <ProjectReference Include="..\testproject\TestProject.AppHost\TestProject.AppHost.csproj" />
-    
+    <ProjectReference Include="..\..\src\Aspire.Hosting\Aspire.Hosting.csproj" AppHostLibrary="true" />
```
Why is the test project configured with `IsAspireHost=true`, thus necessitating all the `AppHostLibrary` overrides?

eerhardt replied:
IsAspireHost was added in https://github.com/dotnet/aspire/pull/560. I assume it is because the Hosting tests create DistributedApplicationBuilder instances inline in the tests. @davidfowl ?

davidfowl replied:
Correct.

mitchdenny commented on playground/eShopLite/AppHost/AppHost.csproj:
```diff
@@ -10,9 +10,10 @@
-    <ProjectReference Include="..\..\..\src\Aspire.Hosting\Aspire.Hosting.csproj" />
-    <ProjectReference Include="..\..\..\src\Aspire.Hosting.Azure\Aspire.Hosting.Azure.csproj" />
-    <ProjectReference Include="..\..\..\src\Aspire.Hosting.Azure.Provisioning\Aspire.Hosting.Azure.Provisioning.csproj" />
+    <ProjectReference Include="..\..\..\src\Aspire.Hosting\Aspire.Hosting.csproj" AppHostLibrary="true" />
```
So this will be required anytime you want to add a reference to a library and not treat it like a resource project?

eerhardt replied:
Correct.

@rainersigwald suggested that we could add a build warning if we detect a `<ProjectReference` to a `Library` project and tell you that it won't get referenced by default, suggesting to add `AppHostLibrary="true"` to the ProjectReference. I think that is a good idea and am going to implement it here.

BTW - anyone have a better name than `AppHostLibrary` for this property?

baronfel replied:
How solid is the term `AppHost`? It feels pretty broad - my gut is that the property should be scoped with `Aspire` in some way to narrow it somewhat. Maybe `AspireHostLibrary`? Total gut feel though.

eerhardt replied:
> @rainersigwald suggested that we could add a build warning if we detect a <ProjectReference to a Library project and tell you that it won't get referenced by default, suggesting to add AppHostLibrary="true" to the ProjectReference. I think that is a good idea and am going to implement it here.

I have added this build warning. I needed to write a C# task, but we don't have a task library today to put it in. So I used a RoslynCodeTaskFactory to get unblocked here. Once this is merged I plan on logging an issue to move this code into a proper Tasks .dll. I'm trying to not get blocked by adding that infrastructure here.

eerhardt committed:
Update CosmosEndToEnd app

eerhardt committed:
Rename AppHostLibrary => AspireHostLibrary

radical commented:
> To resolve these issues, we are taking the following approach:
> - By default ProjectReferences in the AppHost will mean a reference to a "service application", and not a regular project reference. If an AppHost project truly wants a ProjectReference to another project, it can be marked as <ProjectReference Include="..." AppHostLibrary="true" />.

Reversing this behavior might make more sense. The default would be a regular `ProjectReference`, as one would expect. And for a *service-application-project-reference* you can set some `ServiceApplicationLibrary="true"`. That would be less surprising.

eerhardt committed:
Add a build warning informing developers when they ProjectReference a library in the AppHost, but didn't set AspireHostLibrary=true.

eerhardt commented:
> Reversing this behavior might make more sense.

That would fall into the same category as approach (1) in https://github.com/dotnet/aspire/issues/1563#issuecomment-1898826093 - adding a new item type like `AspireProjectResource`. The problem with those approaches is that tooling doesn't know how to add a new item, or set this special metadata on `ProjectReference`. We would need to teach the tooling, for example add a new `dotnet add aspire-reference` command.

The 80+% case here is that when an AppHost has a `ProjectReference`, it is to an executable that it wants to start when the AppHost starts. That's why I think using an empty `ProjectReference` for this case makes the most sense.

danmoseley commented on src/Aspire.Hosting/build/Aspire.Hosting.targets:
```diff
@@ -104,14 +98,83 @@ internal static class ]]>%(ClassName)<![CDATA[
+    <Task>
+      <Using Namespace="System.Xml.Linq" />
+      <Code Type="Fragment" Language="cs">
+        <![CDATA[
```
@rainersigwald  what's the perf cost of inline tasks these days? Originally they used the old CodeDOM and were super slow to compile and not cached.

eerhardt replied:
On my machine, when it was “cold” with no node reuse, the task took 700ms. On a warm incremental build it took 70ms.

I plan on logging an issue to move this task into a .dll, but didn’t want to block on that infrastructure.

danmoseley replied:
Cool. Maybe they do it in memory now, or cache.

eerhardt replied:
Logged https://github.com/dotnet/aspire/issues/1825

eerhardt committed:
Add test to ensure warning is emitted.

joperezr commented on playground/CosmosEndToEnd/CosmosEndToEnd.AppHost/CosmosEndToEnd.AppHost.csproj:
```diff
@@ -9,8 +9,9 @@
   <ItemGroup>
-    <ProjectReference Include="..\..\..\src\Aspire.Hosting.Azure\Aspire.Hosting.Azure.csproj" />
-    <ProjectReference Include="..\..\..\src\Aspire.Hosting\Aspire.Hosting.csproj" />
+    <ProjectReference Include="..\..\..\src\Aspire.Hosting.Azure\Aspire.Hosting.Azure.csproj" AspireHostLibrary="true" />
```
NIT: Not really a suggestion, but I'm not sure I love the name of the metadata. From a user's POV it is not obvious what this means until you understand the context. I wonder if we should instead use something more explicit, like `AddReferenceToAssembly` or something that is more obvious.

eerhardt replied:
I agree that I'm not super happy about this name, but I haven't found one I like better yet.

The metadata does more than just take away the reference to the assembly, it is really the switch between "is this an app that you want to orchestrate?" or "is this a library that has AppHost extension logic you want to use?".

Another idea is to flip it to be `AspireServiceApp=false`. We already have `ServiceNameOverride` if you want to specify the name of the `Projects` class. Setting `AspireServiceApp=false` would mean "this isn't an application, so reference it like a normal ProjectReference".

radical replied:
super nit: Prefixing the name with `Is*` would make it more readable.

eerhardt replied:
> super nit: Prefixing the name with `Is*` would make it more readable.

Agreed. I think initially I didn't add `Is*` because the other ProjectReference metadata didn't follow this pattern:

* ReferenceOutputAssembly 
* SkipGetTargetFrameworkProperties 
* ExcludeAssets 

But these are more "should" and not "is", so I think it is a different scenario.

So should we have:

* `IsAspireHostLibrary=true`
* -or-
* `IsAspireServiceApp=false`
?

I think I'm leaning towards `IsAspireHostLibrary=true` because the naming aligns with `IsAspireHost=true` for the whole project.

radical replied:
I don't really know yet all the special stuff done for these projects, so not sure what a `ServiceApp` vs `Library` would mean.

joperezr replied:
ServiceApp is literally any type of project which represents a microservice that you want to launch via the AppHost. This means the AppHost will orchestrate those types of projects.

The AppHost, while it is in reality an orchestrator, in the end is just a C# project, so you could think of scenarios where people want to add project references to libraries too (which aren't really microservices). For example, imagine a company that has several Aspire apps, and they want to share some of the logic they have in their AppHost projects. For this, they would build a shared library, which AppHost projects would all reference. This library, is not really a service, but instead just a library used by the AppHost.

radical replied:
So:
- `ServiceApp` - special project that the AppHost project will *use* in some way, maybe post-process etc.
- Other case - simple *library* projects, and AppHost will reference the assemblies from these, and use them
- Wasn't there another case of referencing projects that are simply meant as a build dependency, and should not affect AppHost project's build itself?

eerhardt replied:
> ServiceApp - special project that the AppHost project will use in some way, maybe post-process etc.

It isn't a "special" project. But ServiceApps are just regular ASP.NET apps - like a web api, or blazor server app. They can also be "Worker" apps, if they don't need to expose an HTTP endpoint.

The AppHost's Program.cs files uses these apps by calling `builder.AppProject<ServiceApp>("serviceName")`.

> Wasn't there another case of referencing projects that are simply meant as a build dependency, and should not affect AppHost project's build itself?

I don't know of this case. In my head there are mainly 2 cases: ServiceApps and "normal" ProjectReferences to libraries.

radical replied:
> I don't know of this case. In my head there are mainly 2 cases: ServiceApps and "normal" ProjectReferences to libraries.

My misunderstanding then.

I would vote for `IsAspireServiceApp=false` because we are changing the default `ProjectReference` to mean a ServiceApp reference which once you know it, makes sense. IOW, `AppHost` project references `ServiceApp`s by default. In that case for any other project references an opt-out of that default would make sense.

eerhardt replied:
Discussing this more with @DamianEdwards, we decided to go with the name:

* `IsAspireProjectResource=false`
    * Setting this to false opts you into being an "AppHost library", you don't get "Projects" code generated, and you start referencing the project

to align with the "Resource" terminology in the app model.

We can then do some more renames:

* `ServiceNameOverride` => `AspireProjectMetadataTypeName`
* `IServiceMetadata` => `IProjectMetadata`

joperezr commented on src/Aspire.Hosting.Sdk/SDK/Sdk.targets:
```diff
@@ -20,7 +18,21 @@
+      <_IsServiceApp Condition="'%(_IsServiceApp)' == '' and '%(AspireHostLibrary)' != 'true'">true</_IsServiceApp>
+
+      <ReferenceOutputAssembly Condition="'%(ReferenceOutputAssembly)' == '' and '%(_IsServiceApp)' == 'true'">false</ReferenceOutputAssembly>
+      <SkipGetTargetFrameworkProperties Condition="'%(SkipGetTargetFrameworkProperties)' == '' and '%(_IsServiceApp)' == 'true'">true</SkipGetTargetFrameworkProperties>
```
I'm assuming this works transitively, right? Like if AppHost references a project which then references another project?

eerhardt replied:
Correct - that's the intention. This is why I needed to put this logic in the Sdk.targets file, so NuGet Restore doesn't transitively add references of the referenced projects.

joperezr commented on src/Aspire.Hosting.Sdk/SDK/Sdk.targets:
```diff
@@ -37,6 +49,6 @@
   we can move it to WorkloadManifest.targets to ensure future correctness.
   -->
-  <Import Project="Sdk.targets" Sdk="Aspire.Hosting.Orchestration" />
+  <Import Project="Sdk.targets" Sdk="Aspire.Hosting.Orchestration" Condition="'$(_SuppressImportAspireHostingOrchestrationTargets)' != 'true'" />
```
I don't love this change, as we are now leaking an implementation detail of how we build and test stuff in the repo into our shipped SDK. That said, I'd have to play a bit more with this to see if there are other alternatives. Why wasn't this needed before?

eerhardt replied:
We never imported the Sdk.targets before. But with the transitively change mentioned above, we now need to import the Sdk.targets file from our AppHost projects in the repo (and the new test I added). Importing the `Aspire.Hosting.Orchestration` targets doesn't work here, which is why I added the opt out mechanism.

This is roughly equivalent to what we have in `SkipAspireWorkloadManifest`:

https://github.com/dotnet/aspire/blob/0d1f8ea7750b69c8f221816077e9e5a5baa1bc0a/src/Microsoft.NET.Sdk.Aspire/WorkloadManifest.in.targets#L3-L6

joperezr replied:
fair enough.

eerhardt committed:
Skip checking for the Aspire workload in the test

radical commented on tests/Aspire.Hosting.Tests/Aspire.Hosting.Tests.csproj:
```diff
@@ -8,11 +8,18 @@
+    <ProjectReference Include="..\..\src\Components\Aspire.Npgsql\Aspire.Npgsql.csproj" AspireHostLibrary="true" />
+    <ProjectReference Include="..\..\src\Components\Aspire.Oracle.EntityFrameworkCore\Aspire.Oracle.EntityFrameworkCore.csproj" AspireHostLibrary="true" />
+
     <PackageReference Include="Microsoft.Extensions.Http.Resilience" />
```
Should this have `Version="$(MicrosoftExtensionsHttpResiliencePackageVersion)"`. And similarly a fixed version for `Polly` too?

eerhardt replied:
We use CPM (central package management) in Directory.Packages.props in the root of the repo.

radical replied:
ah, thank you!

joperezr commented on src/Aspire.Hosting/build/Aspire.Hosting.targets:
```diff
@@ -12,18 +12,22 @@
-          DependsOnTargets="FindReferenceAssembliesForReferences"
-          Inputs="@(ReferencePathWithRefAssemblies)"
-          Outputs="@(ReferencePathWithRefAssemblies)">
+          DependsOnTargets="_CreateServiceAppProjects">
```
NIT: Not really added in your changes, but I noticed this target is not private (doesn't start with "_") but the rest are, is there a special reasoning behind that?

eerhardt replied:
The reasoning is to support other languages (i.e. F#). See the comment here:

https://github.com/dotnet/aspire/blob/e344c7c1556bf7f52163ab6fe3ac94ef0c395835/src/Aspire.Hosting/build/Aspire.Hosting.targets#L108

New languages could plug-in to `WriteServiceMetadataSourcesDependsOn` (which I made "public" in this change), and then be `DependsOn="CreateServiceMetadataSources"`, and then generate the code for the language using the `ServiceMetadataSource` item.

joperezr commented on src/Aspire.Hosting/build/Aspire.Hosting.targets:
```diff
@@ -32,33 +36,27 @@
-[System.Diagnostics.DebuggerDisplay("Type = {GetType().Name,nq}, ProjectPath = {ProjectPath}")]
-public class ]]>%(ClassName)<![CDATA[ : IServiceMetadata
+[global::System.Diagnostics.DebuggerDisplay("Type = {GetType().Name,nq}, ProjectPath = {ProjectPath}")]
+public class ]]>%(ClassName)<![CDATA[ : global::Aspire.Hosting.IServiceMetadata
```
NIT: Should we also generate an `ExcludeFromCodeCoverage` attribute? or altenarively a `Generated` one?

eerhardt replied:
I've opened https://github.com/dotnet/aspire/issues/1792 for this.

joperezr commented on src/Aspire.Hosting/build/Aspire.Hosting.targets:
```diff
@@ -104,14 +98,83 @@ internal static class ]]>%(ClassName)<![CDATA[
+      <Output TaskParameter="NonExecutableReferences" ItemName="_NonExecutableServiceApps" />
+    </GetNonExecutableReferences>
+
+    <Warning Code="ASPIRE004"
```
Should this be an error instead? Will there ever be the case where someone references a non-exe and it doesn't want to reference the real assembly? I think even for IHostedServices that won't be the case.

eerhardt replied:
I think going to an error would be too far. You can shut this behavior off, which typically isn't possible for "errors". It isn't typically an error that we can't continue. It is just something you probably didn't intend.

joperezr replied:
okay, fair enough.

joperezr commented on src/Aspire.Hosting/build/Aspire.Hosting.targets:
```diff
@@ -104,14 +98,83 @@ internal static class ]]>%(ClassName)<![CDATA[
+  Validates that all the ProjectReferences of an Aspire AppHost project are executables and
+  informs the developer to set 'AspireHostLibrary=true' if they really intended on ProjectReferencing a library.
+  -->
+  <Target Name="_ValidateAspireHostServiceReferences"
```
While I'm totally fine with this approach, I wonder if we also evaluated the alternative approach, where the projects that require the additional metadata are the service projects, instead of the libraries. Here is my thinking: we expect most of the service project references to be added via VS, by clicking the "Enlist project to Aspire Host" item, so we can easily code that to set the additional metadata. On the other hand, it is likely that when adding a project reference to a library, this will be done manually, and therefore will today have to know to set the metadata. I wonder if we should try to optimize for having the manual case (which in this message I'm assumig it will be libraries) not require metadata.

Thoughts?

joperezr replied:
Ok, nevermind, looks like this was already asked and discussed above. CLosing.

eerhardt replied:
Answered in https://github.com/dotnet/aspire/pull/1726#issuecomment-1905065215.

joperezr commented on tests/Aspire.Hosting.Tests/MSBuildTests.cs:
```diff
@@ -0,0 +1,136 @@
+            var outputDone = new ManualResetEvent(false);
+            var process = new Process();
+            // set '-nodereuse:false -p:UseSharedCompilation=false' so the MSBuild and Roslyn server processes don't hang around, which may hang the test in CI
+            process.StartInfo = new ProcessStartInfo("dotnet", $"build -nodereuse:false -p:UseSharedCompilation=false")
```
The only thing that worries me from this is how dotnet resolution might differ in different machines and CI. I wonder if we should get .dotnet by analyzing the CurrentProcess arg0, so that we use the same dotnet as the one that was launched for running these tests.

radical replied:
For running tests on the build machine setting that or explicitly using `dotnet` from `.dotnet` should be enough.
On helix there will only be the `dotnet` with version matching the one in `global.json`, so it shouldn't be an issue there.

radical replied:
Local runs can do the same thing - compute a path to `.dotnet/dotnet`, and use that.

joperezr replied:
> or explicitly using dotnet from .dotnet should be enough.

I don't think we can rely on this. If the  machine already has the SDK listed in global.json globally installed, then the build will skip restoring it and creating a .dotnet folder, and instead will just use the global one. We've been bit by this in the past.

radical replied:
Good point. I guess in that case use `dotnet` from `PATH`, which then should be the correct version?

joperezr replied:
@eerhardt can correct me if I'm wrong, but I believe you can get to the one running the tests via Process.GetCurrentProcess() and then inspecting arg0

danmoseley replied:
I think that only works when you set it on the ProcessStartInfo in the first place and used that to start the process?

alternative: Process.GetCurrentProcess().MainModule.FileName;

eerhardt replied:
These approaches don't work because the process being run is:

> C:\git\aspire\artifacts\bin\Aspire.Hosting.Tests\Debug\net8.0\testhost.exe

We are already getting the "repo root" in this test. So the only viable options I see are:

1. Doing what I'm doing here and using the `dotnet` on the path.
2. Using `{repoRoot}\.dotnet\dotnet`.

Thoughts on those 2 options? (or are there others?)

radical replied:
You can't depend on (2) because of the reason @joperezr mentioned where your system dotnet version matches the one in global.json, thus you don't get any `.dotnet`.

One possible thing is to add `$(DotnetTool)` from msbuild to `PATH` when running the tests, which should make the expected `dotnet` available in `PATH.

eerhardt replied:
> One possible thing is to add $(DotnetTool) from msbuild to PATH when running the tests

How would one do that? I'm running the tests in VS using the Test Explorer.

Either way, this wouldn't change the test code - we would just invoke `dotnet` and let the PATH take care of it.

radical replied:
forgot about VS! 😬

danmoseley commented on tests/Aspire.Hosting.Tests/MSBuildTests.cs:
```diff
@@ -0,0 +1,138 @@
+
+            var output = new StringBuilder();
+            var outputDone = new ManualResetEvent(false);
+            var process = new Process();
```
nit: not disposed

danmoseley commented on tests/Aspire.Hosting.Tests/MSBuildTests.cs:
```diff
@@ -0,0 +1,138 @@
+                CreateNoWindow = true,
+                WorkingDirectory = appHostDirectory
+            };
+            process.OutputDataReceived += (sender, e) =>
```
might it make sense to subscribe to ErrorDataReceived in case of some error? without introducing the [famous deadlock](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.process.standarderror)

danmoseley replied:
basically just add `output.AppendLine(process.StandardError.ReadToEnd());`  just before WaitForExit.

eerhardt replied:
MSBuild doesn't write to stderr. See https://github.com/dotnet/msbuild/issues/3101.

I think we can add this in the future if it does start writing to stderr.

joperezr commented:
Overall comment: Correct me if I'm wrong, but I don't believe this change is braking for most of the existing Aspire apps, right? meaning, for all of the customers that created Aspire apps using preview2, and will now want to upgrade their workload and references to packages, they won't need additional changes for this to work. (unless of course they are using libraries from their apphost)

eerhardt commented:
> Overall comment: Correct me if I'm wrong, but I don't believe this change is braking for most of the existing Aspire apps, right? meaning, for all of the customers that created Aspire apps using preview2, and will now want to upgrade their workload and references to packages, they won't need additional changes for this to work. (unless of course they are using libraries from their apphost)

Correct this is not breaking for normal Aspire apps. It is breaking for people using ProjectReferences and needing to reference those projects from their AppHost. In that case, they will need to set `AspireHostLibrary=true`.

eerhardt committed:
PR feedback

- Rename AppHostLibrary to IsAppHostLibrary
- Dispose of Process in test

eerhardt committed:
Merge remote-tracking branch 'upstream/main' into FixProjectReference

eerhardt committed:
Fix up merge

eerhardt commented:
I plan on merging this today in order to get it in for preview3 with enough bake time. Let me know if you have any feedback that would block merging this. We can tweak it as we go after merge as well.

eerhardt committed:
Rename ProjectReference properties

- IsAspireProjectResource=false
    - Setting this to false opts you into being an "AppHost library", you don't get "Projects" code generated, and you start referencing the project
- ServiceNameOverride => AspireProjectMetadataTypeName

This also allows us to rename IServiceMetadata to IProjectMetadata.

eerhardt committed:
Fix broken test
